readium/navigator/src/main/java/org/readium/r2/navigator/media2/SessionPlayerFacade.kt

```diff
@@ -0,0 +1,200 @@
+ * Compound commands have two limitations due to the framework weaknesses:
+ * - in case of failure, the playback can be left in an intermediate state
+ * - the behaviour is undefined if any external controller takes actions at the same time
+ */
+@ExperimentalAudiobook
+@OptIn(ExperimentalTime::class)
+internal class SessionPlayerFacade(
+    private val sessionPlayer: SessionPlayer,
+) {
+    val playerState: SessionPlayerState
+        get() = SessionPlayerState.from(sessionPlayer.playerState)
+
+    val currentPosition: Duration?
+        get() = sessionPlayer.currentPositionDuration
+
+    val currentIndex: Int?
+        get() = sessionPlayer.currentMediaItemIndex
+            .takeUnless { it == SessionPlayer.INVALID_ITEM_INDEX }
+
+    val currentDuration: Duration?
+        get() = sessionPlayer.duration
+            .takeUnless { it == SessionPlayer.UNKNOWN_TIME }
+            ?.milliseconds
+
+    val playlist: List<MediaItem>?
+        get() = sessionPlayer.playlist
+
+    val playbackSpeed: Double?
+        get() = sessionPlayer.playbackSpeedNullable
+
+    fun unregisterPlayerCallback(callback: SessionPlayer.PlayerCallback) {
+        sessionPlayer.unregisterPlayerCallback(callback)
+    }
+
+    fun close() {
+        sessionPlayer.close()
+    }
+
+    fun session(context: Context, id: String, activityIntent: PendingIntent): MediaSession {
+        return MediaSession.Builder(context, sessionPlayer)
+            .setId(id)
+            .setSessionActivity(activityIntent)
+            .build()
+    }
+
+    /*
+     * Basic commands
+     */
+
+    suspend fun prepare(): SessionPlayerResult =
+        submitCommand(this::prepareSync)
+
+    suspend fun setPlaylist(playlist: List<MediaItem>, metadata: MediaMetadata): SessionPlayerResult =
+        submitCommand { setPlaylistSync(playlist, metadata) }
+
+    suspend fun play(): SessionPlayerResult =
+        submitCommand(this::playSync)
+
+    suspend fun pause(): SessionPlayerResult =
+        submitCommand(this::pauseSync)
+
+    suspend fun setPlaybackSpeed(speed: Double): SessionPlayerResult =
+        submitCommand { setPlaybackSpeedSync(speed) }
+
+    /*
+     * Compound commands
+     */
+
+    suspend fun seekTo(index: Int, position: Duration): SessionPlayerResult =
+        submitCommandOnPause {
+            when {
+                index == sessionPlayer.currentMediaItemIndex -> {
+                    seekToSync(position)
+                }
+                position == Duration.ZERO -> {
+                    skipToPlaylistItemSync(index)
+                }
+                else -> {
+                    skipToPlaylistItemSync(index)
+                        // WORKAROUND: skipToPlayListItem terminates before Exoplayer is aware that
+                        // the new item is seekable. Not to wait would result in the command being skipped.
+                        .also { Thread.sleep(500) }
```

Comment:
Isn't there a way to make this a bit less fragile? There's no callback event when the current item changes that could be used?

Reply:
No, I think that's a bug. Looking at the code, if I remember correctly, I supposed the callback would be called too early as well.